src/analyze-dependencies.ts

```diff
@@ -117,70 +118,39 @@ export class LocalDependencyAnalyzer implements DependencyAnalyzer {
     seenPackages = new Set<string>()
   ) {
     try {
-      const entries = await fs.readdir(dir, {withFileTypes: true});
-
-      for (const entry of entries) {
-        const fullPath = path.join(dir, entry.name);
-
-        // Handle symlinks
-        if (entry.isSymbolicLink()) {
-          try {
-            const realPath = await fs.realpath(fullPath);
-            logger.debug('Found symlink:', fullPath, '->', realPath);
-            // If the real path is a package, process it
-            const pkgJsonPath = path.join(realPath, 'package.json');
-            try {
-              const pkgJson = JSON.parse(
-                await fs.readFile(pkgJsonPath, 'utf-8')
-              );
-              if (!seenPackages.has(pkgJson.name)) {
-                seenPackages.add(pkgJson.name);
-                logger.debug('Detected package (symlink):', pkgJson.name, 'at', realPath);
-                callbacks.onPackage(pkgJson);
-              } else {
-                logger.debug('Already seen package (symlink):', pkgJson.name, 'at', realPath);
-              }
-            } catch {
-              // Not a package or can't read package.json, continue
-            }
-            // Only follow symlinks that point to node_modules
-            if (realPath.includes('node_modules')) {
-              logger.debug('Following symlink to:', realPath);
-              await this.walkNodeModules(realPath, callbacks, seenPackages);
-            }
-          } catch {
-            logger.debug('Error resolving symlink:', fullPath);
-          }
-          continue;
-        }
-
-        if (entry.isDirectory()) {
-          // Check if this is a package directory
-          const pkgJsonPath = path.join(fullPath, 'package.json');
+      const crawler = new fdir()
+        .withFullPaths()
+        .withSymlinks()
+        .filter((filePath) => {
+          // Skip node_modules directories to avoid infinite recursion
+          return !filePath.includes('/node_modules/node_modules/');
```

Comment:
is this meant to say `/node_modules` once?

Reply:
oh, it didn't happen to me, but it's a common defensive practice when traversing directories to prevent infinite loops. i added it as a precaution to avoid potential infinite recursion in case of a nested dir or a symlink pointing to one exists. while it's rare, it can happen in certain scenarios e.g. if a pkg's installation process creates a symlink that points back to the top lvl `node_modules` or if a nested dep is installed in a way that creates a circular ref.

in the end if you think that such scenario won't happen to us, i could simplify the filter.

Reply:
it seems unlikely that'd be `node_modules/node_modules` though. more like `node_modules/something/node_modules`

maybe we're overthinking here. why cover `node_modules/node_modules` but not `node_modules/foo/node_modules`? or any other infinite symlink i can make

i think we should either detect recursion generically or not at all

Reply:
would something like this work?

```ts
          // Get the directory part of the path
          const dirPath = path.dirname(filePath);
          // If we've seen this directory before, it's a cycle
          if (visitedPaths.has(dirPath)) {
            logger.debug('Detected cycle at:', dirPath);
            return false;
          }
          return true;
        })
```